torch/nn/parallel/distributed.py

```diff
@@ -441,7 +449,7 @@ def forward(self, *inputs, **kwargs):
         if self.require_forward_param_sync:
             self._sync_params()
 
-        if self.device_ids:
+        if self.device_ids and (len(self.device_ids) > 1 or self.scatter_single_device):
```

Comment:
I think #22384 failed because the input was not copied from CPU to GPU. Could you please try if the following would work?

```python

# add a internal helper function
def to(self, obj, device):
    if isinstance(obj, torch.Tensor):
        return obj.to(device)
    if isinstance(obj, tuple) and len(obj) > 0:
        return list(zip(*map(self.to, obj)))
    if isinstance(obj, list) and len(obj) > 0:
        return list(map(list, zip(*map(self.to, obj))))
    if isinstance(obj, dict) and len(obj) > 0:
        return list(map(type(obj), zip(*map(self.to, obj.items()))))
    return obj

# add this to the above line in forward function
if self.device_ids:
    if len(self.device_ids) == 1:
        inputs = self.to(inputs, self.device_ids[0])
        kwags = self.to(kwargs, self.deivce_ids[0])
        output = self.module(*inputs[0], **kwargs[0])
    else:
        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
        outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
        output = self.gather(outputs, self.output_device)
else:
    ...
```

The reason for the above suggestion is that I am a little hesitate to expose a flag for a very specific use case, and it seems this can be addressed without that flag.

@gchanan Do we have any API similar to the `self.to(obj, device)` above that we can use here?

Reply:
Thanks for the helper function, but this will also break the (possibly arbitrarily nested) data structure in my case.  I have a similar helper function intended to work for either tuples or namedtuples (and our custom namedarraytuples), which have a different construction style:

```
def buffer_to(buffer_, device=None):
    if buffer_ is None:
        return
    if isinstance(buffer_, torch.Tensor):
        return buffer_.to(device)
    elif isinstance(buffer_, np.ndarray):
        raise TypeError("Cannot move numpy array to device.")
    contents = tuple(buffer_to(b, device=device) for b in buffer_)
    if type(buffer_) is tuple:
        return contents
    return type(buffer_)(*contents)
```

Not sure if you would want to support something so specific, probably not.

But I don't understand why the DDP model should be responsible for moving inputs to the GPU when it's only running a single device, that's not consistent with the usual model, which expects data to already be on the right device.  Would prefer to bypass this intermediary processing.

Reply:
> Not sure if you would want to support something so specific, probably not.

Right, this is too specific to fit in DDP. And I see why it the `to` helper I posted above wouldn't work.

> But I don't understand why the DDP model should be responsible for moving inputs to the GPU when it's only running a single device, that's not consistent with the usual model, which expects data to already be on the right device.

This is a good point. I think the motivation might be keep it consistent with `DataParallel` which does scatter inputs and args.

Reply:
I'm not aware of any function like that that exists, although I don't see anything Pytorch / "to" specific about it